src/base.jl

```diff
@@ -48,21 +46,65 @@ size_nibbles(::Type{T}) where T = ceil(log2(sizeof(T))/4)
 
 size_mask(T) = UInt(exp2(4*size_nibbles(T)) - 1)
 
+#==
+function Base.getindex(s::ShortString, i::Integer)
+     getindex(String(s), i)
+end
 
-# function Base.getindex(s::ShortString, i::Integer)
-#     getindex(String(s), i)
-# end
+function Base.getindex(s::ShortString, args...; kwargs...)
+     getindex(String(s), args...; kwargs...)
+end
+==#
+
+@inline function length(s::ShortString{T}) where T
+    i = 0
+    len = 0
+    while i < ncodeunits(s)
+        shifted = s.size_content >> (8*(sizeof(T) - i))
+        i += if shifted % UInt8 <= 0x7f  # 1 byte character
+            1
+        elseif shifted % UInt16 <= 0x7ff  # 2 byte character
+            2
+        elseif shifted % UInt32 <= 0xffff  # 3 byte character
+            3
+        else  # 4 byte character
+            4
+        end
+        len += 1
+    end
+    return len
+end
 
-# function Base.getindex(s::ShortString, args...; kwargs...)
-#     getindex(String(s), args...; kwargs...)
-# end
+@inline function Base.iterate(s::ShortString{T}, i::Integer=1) where T
+    i > ncodeunits(s) && return nothing
+    shifted = s.size_content >> (8*(sizeof(T) - i))
+    if shifted % UInt8 <= 0x7f  # 1 byte character
+        return Char(shifted % UInt8), i+1
+    elseif shifted % UInt16 <= 0x7ff  # 2 byte character
+        return Char(shifted % UInt16 % UInt32), i+2
+    elseif shifted % UInt32 <= 0xffff  # 3 byte character
+        return Char(shifted % UInt32 & 0xffff), i+3
+    else  # 4 byte character
+        return Char(shifted % UInt32), i+4
+    end
+end
 
-Base.collect(s::ShortString) = collect(String(s))
+function ==(s::ShortString{S}, b::AbstractString) where S
+    ncodeunits(b) == ncodeunits(s) || return false
+    return s == ShortString{S}(b)
+end
+==(a::AbstractString, b::ShortString) = b == a
+function ==(a::ShortString{S}, b::ShortString{S}) where S
+    return a.size_content == b.size_content
+end
 
-==(s::ShortString, b::AbstractString) = begin
-    String(s) == b
+function Base.hash(s::ShortString, h::UInt)
+    h += Base.memhash_seed
+    mmhash128_c(s, h % UInt32)[2] + h
```

Comment:
Doing as I suggest will also totally avoid allocations.  This can be important for keeping the speed of using these in things like Dicts fast.

Reply:
Maybe you can make a seperate PRs for that?
It is beyond me.
The internals of MurmurHash3 scare me.
Also it would not depend on having iterate defined so it would be able to be done seperate form this PR

Reply:
> Maybe you can make a seperate PRs for that?

I would be keen to learn from that PR actually.

@oxinabox @ScottPJones do any of you want to take over ShortStrings.jl? Cos I only made it to make the experiment group by faster for strings. And strings were much slower in the past. Now radixsort for strings is ok in SortingLab.jl I barely touch this repo.

Reply:
This looks like other people very well may want to use it, so please consider moving it to JuliaString 😄